### PR TITLE
Server-api improvements, support plugin development in TypeScript

### LINKED ISSUE #1516: Interfaces are tightly coupled to signalk-server
The interface required for developing a plugin in typescript is buried in the signalk-server package. 

Steps to reproduce:

Create a separate plugin GitHub project, and implement typescript.
When exporting the default method as specified in the plugin documentation, include the app parameter and interface.

`
import { ServerAPI } from 'signalk-server/lib/interfaces/plugins';

module.exports = (app: ServerAPI) => { 
...
}
`

Solution: abstract public interfaces into a type definitions

### PR DESCRIPTION
**Pull request by tkurki**
- Move base Signal K types to server-api so that both server and ts plugins can use them
- Move ServerAPI to server-api module (fixes #1516)

### CODE REVIEW

**Review by panaaj — APPROVED** [review]

File: packages/server-api/src/index.ts [hunk 1]
```diff
@@ -89,3 +89,75 @@ export interface Plugin {
...
+  //TSTODO convert queryRequest to ts
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  queryRequest: (requestId: string) => Promise<any>
+  error: (msg: string) => void
+  debug: (msg: string) => void
+  registerDeltaInputHandler: (handler: DeltaInputHandler) => void
+  setProviderStatus: (msg: string) => void
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  handleMessage: (id: string, msg: any, skVersion?: SKVersion) => void
+  setProviderError: (msg: string) => void
```
Comment (panaaj) [inline packages/server-api/src/index.ts]: Hasn't `setProviderStatus` & `setProviderError`been deprecated and replaced by `setPluginStatus` and `setPluginError`?
Or have I got that the wrong way around?

**Review by panaaj — COMMENTED** [review]
The `app` object passed to the plugin also has a `config` property which has properties like `configPath` which can be useful.
Definitions for the appropriate properties should be considered.

File: packages/server-api/src/index.ts (hunk 1 above)
Comment (tkurki) [reply packages/server-api/src/index.ts]: No, that's correct. Force pushed a fix.

**Comment by tkurki** [comment]
https://github.com/SignalK/signalk-server/blob/c08d12b9882b293e9f7b71e44ba91ba9c7735548/src/config/config.ts#L38-L79

**Comment by tkurki** [comment]
To be honest I don't think we should expose the whole config to plugins, even if things are so today. Is there something specific there that you'd like to see exposed?

In fact I would like to turn the tables on the app object: not expose everything by default (and override some of the things) and instead only expose read only copies of the originals, defining the server api for plugins in a much more controlled way.

This PR has already a bunch of stuff that is very usable for a "normal" ts plugin that does not reach further into the App object, so I am leaning towards merging and publishing this instead of piling more onto this PR.

Thoughts?

**Comment by panaaj** [comment]
> To be honest I don't think we should expose the whole config to plugins, even if things are so today. Is there something specific there that you'd like to see exposed?

The resources-provider uses  `app.config.configPath` as the default path for creating folders to store resources.
This can be changed by the user but I provides a suitable default.

If the plugin wants to query the server "settings" is there another means of doing it?
I think it would be useful for plugins to have read-only access to this type of information.

**Comment by panaaj** [comment]
> In fact I would like to turn the tables on the app object: not expose everything by default (and override some of the things) and instead only expose read only copies of the originals, defining the server api for plugins in a much more controlled way.

Sure, I think that is a good approach, as long as plugins have access to the required functions and properties.

**Comment by panaaj** [comment]
> This PR has already a bunch of stuff that is very usable for a "normal" ts plugin that does not reach further into the App object, so I am leaning towards merging and publishing this instead of piling more onto this PR.

I have no problem with that.

**Comment by panaaj** [comment]
@tkurki Just adding to the `config` point, I think it would be useful if the plugin had an interface to determine the: protocol, host & port
that the server is operating with.
`app.config.getExternalPort()` does not seem to return anything in my configuration (I would expect it to return 3000)
`app.config.getExternalHostname()` does return a value.

This need has come up when adding PMTiles support to the `charts-plugin`.... to read information from the file a url needs to be supplied (rather than the file path) so the plugin needs to add protocol,host & port for files hosted on the local machine so the metadata can be read.

This information is available from an http request but it inthis instance it is required on plugin startup.

**Comment by panaaj** [comment]
@tkurki Just adding to the `config` point, I think it would be useful if the plugin had an interface to determine the: protocol, host & port
that the server is operating with.
`app.config.getExternalPort()` does not seem to return anything in my configuration (I would expect it to return 3000)
`app.config.getExternalHostname()` does return a value.

This need has come up when adding PMTiles support to the `charts-plugin`.... to read information from the file a url needs to be supplied (rather than the file path) so the plugin needs to add protocol,host & port for files hosted on the local machine so the metadata can be read.

This information is available from an http request but it inthis instance it is required on plugin startup.

**Comment by tkurki** [comment]
🤦 we have `getExternalPort` in both ports.ts and config.ts and they behave *ahem* differently. Feel like fixing that? And when fixed we could add related getters to the server api for plugins: getExternalPort, getExternalHost and I think getSsl while we are at it.

**Comment by tkurki** [comment]
As resources provider is a plugin why not use https://github.com/SignalK/signalk-server/blob/master/SERVERPLUGINS.md#appgetdatadirpath ? Principle of least surprise would make me search there for the data if I needed that.

**Comment by panaaj** [comment]
> we have getExternalPort in both ports.ts and config.ts and they behave ahem differently. Feel like fixing that? 

Sure.
Based on what I've seen I feel the definition in config.ts is the one to use moving forward.

File: packages/server-api/src/index.ts [hunk 2]
```diff
@@ -87,3 +89,75 @@ export interface Plugin {
...
+  byPath: string[]
+  byOpenPlotter: string[]
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  serialports: any
+}
+
+export interface ServerAPI extends PluginServerApp {
+  getSelfPath: (path: string) => void
+  getPath: (path: string) => void
+  getMetadata: (path: string) => void
```
Comment (panaaj) [inline packages/server-api/src/index.ts:109]: These definitions do not have any return type.

Comment (tkurki) [reply packages/server-api/src/index.ts:109]: Merged already, start a new one please.

**Pull request merged**